components/scream/src/share/grid/remap/horizontal_remap_utility.cpp

```diff
@@ -0,0 +1,543 @@
+  int my_start = 0;
+  for (int ii=0; ii<my_rank;ii++) {
+    my_start += chunks_glob[ii];
+  }
+  // Check that the total set of chunks covers all the data
+  {
+    int chunk_check = 0;
+    for (int ii=0; ii<num_ranks; ii++) {
+      chunk_check += chunks_glob[ii];
+    }
+    EKAT_REQUIRE_MSG(chunk_check==remap_size,"ERROR: HorizontalMap " + m_name +" get_remap_indices - Something went wrong distributing remap data among the MPI ranks");
+  }
+  // Using scream input routines, read remap data from file by chunk
+  view_1d<int>  tgt_col("row",my_chunk); 
+  auto tgt_col_h = Kokkos::create_mirror_view(tgt_col);
+  std::vector<std::string> vec_of_dims = {"n_s"};
+  std::string i_decomp = "int-n_s";
+  scorpio::get_variable(remap_filename, "row", "row", vec_of_dims, "int", i_decomp);
+  std::vector<int64_t> var_dof(my_chunk);
+  std::iota(var_dof.begin(),var_dof.end(),my_start);
+  scorpio::set_dof(remap_filename,"row",var_dof.size(),var_dof.data());
+  scorpio::set_decomp(remap_filename);
+  scorpio::grid_read_data_array(remap_filename,"row",0,tgt_col_h.data(),tgt_col_h.size()); 
+  scorpio::eam_pio_closefile(remap_filename);
+  // Step 2: Now that we have the data distributed among all ranks we organize the data
+  //         into sets of target column, start location in data and length of data.
+  //         At the same time, determine the min_dof for remap column indices.
+  std::vector<int> chunk_dof, chunk_start, chunk_len;
+  chunk_dof.push_back(tgt_col_h(0));
+  chunk_start.push_back(my_start);
+  chunk_len.push_back(1);
+  int remap_min_dof = tgt_col_h(0);
+  for (int ii=1; ii<my_chunk; ii++) {
+    remap_min_dof = std::min(tgt_col_h(ii),remap_min_dof);
+    if (tgt_col_h(ii) == chunk_dof.back()) {
+      // Then we add one to the length for this chunk.
+      chunk_len.back() ++;
+    } else {
+      // Start a new chunk for a new DOF
+      chunk_dof.push_back(tgt_col_h(ii));
+      chunk_start.push_back(my_start+ii);
+      chunk_len.push_back(1);
+    }
+  }
+  // Pass chunk information among all ranks so they can be consolidated.
+  int  num_chunks          = chunk_dof.size();
+  int* num_chunks_per_rank = new int[num_ranks];
+  int* chunk_displacement  = new int[num_ranks];
+  int  total_num_chunks;
+  int  global_remap_min_dof;
+  m_comm.all_gather(&num_chunks, num_chunks_per_rank,1);
+  m_comm.all_reduce(&remap_min_dof,&global_remap_min_dof,1,MPI_MIN);
+  chunk_displacement[0] = 0;
+  total_num_chunks = num_chunks_per_rank[0];
+  for (int ii=1; ii<num_ranks; ii++) {
+    chunk_displacement[ii] = total_num_chunks;
+    total_num_chunks += num_chunks_per_rank[ii];
+  }
+  int* buff_dof = (int*)calloc(total_num_chunks, sizeof(int));
+  int* buff_sta = (int*)calloc(total_num_chunks, sizeof(int));
+  int* buff_len = (int*)calloc(total_num_chunks, sizeof(int));
+  MPI_Allgatherv(chunk_dof.data(),  chunk_dof.size(),MPI_INT,buff_dof,num_chunks_per_rank,chunk_displacement,MPI_INT,m_comm.mpi_comm());
```

Comment:
FYI, `ekat::Comm` has some interfaces that make MPI calls a bit more understandable.

Reply:
I looked into that for `allgather` but if I recall either the `allgatherv` wasn't an option, or I kept hitting an error when I tried the implementation in this context.

Reply:
Ah, I missed the `v` part, sorry.